libmcount/mcount.c

```diff
@@ -870,7 +890,10 @@ enum filter_result mcount_entry_filter_check(struct mcount_thread_data *mtdp, un
 	if (mtdp->filter.out_count > 0)
 		return FILTER_OUT;
 
-	uftrace_match_filter(child, &mcount_triggers, tr);
+	if (!agent_run || agent_guard_filter()) {
```

Comment:
I'm not sure if it's really safe.  Even with the full memory barrier, isn't it possible for concurrent readers and the agent can get the writer marker?

Also, I don't like the fact that it might miss the whole filters when agent makes the change.  Can we do it better?

Reply:
I agree this is not safe enough. There are a few ways we can tackle this.

### Disable tracing when agent is working

Our view on this is that if you are interacting with the agent to tweak tracing parameters, you do this for future events, because chances are the current calls are already over when the agent gets the instructions. So you may well skip instrumentation while the agent makes changes.

We can have the agent disable tracing by itself. This requires synchronizing.

We can have the user disabling tracing, and the agent would fail to apply filters if tracing is enabled. Without locks, this comes down to a grace period.

### Use read/write locks

We can introduce read/write locks, and skip instrumentation when data is read-protected. This requires a read lock in the fast path, which comes with an unwanted performance impact.

As discussed during our meeting in January, we could implement this solution in a separate libmcount build, e.g. `libmcount-sync.so`.

### Else

As for lockless algorithms, I don't think that would be simple to do.

Reply:
I'm not happy with the both solutions.  Maybe we could have a simple RCU style algorithm to atomically update the whole filters and free the old copy a few seconds later.  Thoughts?

Reply:
OK. I thought about RCU and we should give it a try. Having a RCU rbtree with per-node RCU is out of reach IMO (quiet complex, and the code I can find is old. See [1] and [2]). Instead we may copy the whole tree. I'll try that.

[1] https://lists.lttng.org/pipermail/lttng-dev/2017-December/027640.html
[2] https://github.com/philip-w-howard/RP-Red-Black-Tree

Reply:
This is out of topic, but I'm curious: why use rbtrees in the first place? They better perform insertion/deletion than lookup, which is the opposite of what libmcount does. Is this from Ftrace?

Reply:
Yeah, I also meant replacing the whole tree.

The rbtree was needed because it needs to check the address range.  Since the location of `mcount` call varies among functions, it couldn't check a specific offset from the function start.  Maybe we can reconsider a hash table if it only supports dynamic tracing because it can control the location.